finality-aleph/src/party.rs

```diff
@@ -229,144 +231,109 @@ async fn run_aggregator<B, C, BE>(
     }
 }
 
-#[allow(clippy::too_many_arguments)]
-async fn run_session_as_authority<B, C, BE, SC>(
-    node_id: NodeIndex,
-    auth_keystore: AuthorityKeystore,
-    client: Arc<C>,
-    data_network: DataNetwork<NetworkData<B>>,
-    session: Session<AuthorityId, NumberFor<B>>,
-    spawn_handle: SpawnHandle,
-    select_chain: SC,
-    metrics: Option<Metrics<B::Header>>,
-    justification_tx: mpsc::UnboundedSender<JustificationNotification<B>>,
-    exit_rx: futures::channel::oneshot::Receiver<()>,
-) where
+impl<B, C, BE, SC> ConsensusParty<B, C, BE, SC>
+where
     B: Block,
     C: crate::ClientForAleph<B, BE> + Send + Sync + 'static,
     C::Api: aleph_primitives::AlephSessionApi<B, AuthorityId, NumberFor<B>>,
     BE: Backend<B> + 'static,
     SC: SelectChain<B> + 'static,
+    NumberFor<B>: From<u32>,
 {
-    debug!(target: "afa", "Authority task {:?}", session.session_id);
-
-    let (ordered_batch_tx, ordered_batch_rx) = mpsc::unbounded();
+    fn run_session_as_authority(
+        &self,
+        node_id: NodeIndex,
+        data_network: DataNetwork<NetworkData<B>>,
+        session: Session<AuthorityId, NumberFor<B>>,
+        exit_rx: futures::channel::oneshot::Receiver<()>,
+    ) -> impl crate::Future<Output = ()> {
```

Comment:
Why `crate::`?

Reply:
To use the same `Future` crate-wide. But you are right that this looks strange, so added an explicit import instead.

Reply:
I still think it would more natural to import `futures::Future` directly, but you can leave it as is

Reply:
I'm not sure why this is exported in the top level of the crate, but since it is using the version from there sounds better to me. I guess we could set up clearer guidelines what we want to do in such situations (whether we can reexport external types, and what to use if we do).